### ISSUE TITLE
Hybrid OX/SREX crossover

### ISSUE THREAD

**Issue by leonlan** [PyVRP/PyVRP#159]
We should re-implement the OX-SPLIT crossover operator, which is originally used by HGS. https://github.com/N-Wouda/Euro-NeurIPS-2022/commit/2c495c4438b6ae24c38bdbb39a5eff6f8bf6d2d9 is the relevant commit from the EURO-NeurIPS repository.

**Comment by N-Wouda** [comment]
SPLIT, or more generally the 'giant tour' representation, is needed for a bunch of different crossover operators right? If we find a central place to implement SPLIT (perhaps on or near the Individual), we can perhaps quickly support a lot of those other crossover operators as well.

**Comment by leonlan** [comment]
I'm working on this issue. My proposal is as follows:
- `make_giant_tour`: creates a giant tour from the given route.
- `ordered_exchange`: performs an ordered exchange crossover
- `split_giant_tour`: a naive splitting algorithm. Specifically, given a tour, we continue adding customers to the current route until no longer feasible, after which we start a new route. If we are the last available vehicle, we simply add all the remaining customers.
- (optional) `alternating_exchange`: a simple alternating exchange (from our Euro-NeurIPS repo)

Everything above is O(n) so I'll probably implement all of these on the Python side.

I don't have enough time to delve into Vidal's [SPLIT](https://github.com/vidalt/Split-Library) algorithms, which are labeling algorithms/dynamic programming algorithms. A naive split algorithm as proposed above should be sufficient for now I think and we can always introduce the more complex ones if needed.

> If we find a central place to implement SPLIT (perhaps on or near the Individual), we can perhaps quickly support a lot of those other crossover operators as well.

My suggestion is to keep the make/split giant tour utilities in `crossover/`. I think it's very specific to crossover and don't see use cases of giant tours outside of that.

**Comment by leonlan** [comment]
Following the terminology in [Prins 2004](https://www.sciencedirect.com/science/article/abs/pii/S0305054803001588), the naive splitting algorithm can be called "sequential split". For CVRP, Vidal proposed a linear time optimal splitting algorithm which is implemented in HGS-CVRP.

**Comment by N-Wouda** [comment]
We can implement a few of these, that might be good from a literature point of view (we can cite a lot of stuff from recent decades). Do we know if these simple crossover operators outperform SREX for e.g. CVRP?

**Comment by leonlan** [comment]
Nope, I haven't seen any papers that compare crossover operators for CVRP. But the paper by Cybula et al. below does compare different combinations of crossover operators for VRP problems with TW.

- [Cybula et al. (2021) - Effective recombination operators for the family of vehicle routing problems](https://dl.acm.org/doi/abs/10.1145/3449726.3459574)
- [Prins et al. (2014) - Order-first split-second methods for vehicle routing problems: A review](https://www.sciencedirect.com/science/article/pii/S0968090X14000230?casa_token=RyqQHm92uVIAAAAA:0Tlp405pXqQ1xWBnwLO5v6ec-em8X0En5EVCZC38lcz536lIo_qjGsYplFiSxN7DuWgqsuYvcoMR#s0010): this may also be relevant to implement other split algorithms

**Comment by wouterkool** [comment]
Indeed the exact SPLIT implementation is here: https://github.com/vidalt/HGS-CVRP/blob/main/Program/Split.cpp or here: https://github.com/ortec/euro-neurips-vrp-2022-quickstart/blob/main/baselines/hgs_vrptw/Split.cpp. I think this has `O(n)` complexity and is optimal so I don't see much benefit of implementing a heuristic in Python which will be both slower and worse (and more work)?

**Comment by leonlan** [comment]
> Indeed the exact SPLIT implementation is here: [vidalt/HGS-CVRP@`main`/Program/Split.cpp](https://github.com/vidalt/HGS-CVRP/blob/main/Program/Split.cpp?rgh-link-date=2023-06-19T06%3A46%3A57Z) or here: [ortec/euro-neurips-vrp-2022-quickstart@`main`/baselines/hgs_vrptw/Split.cpp](https://github.com/ortec/euro-neurips-vrp-2022-quickstart/blob/main/baselines/hgs_vrptw/Split.cpp?rgh-link-date=2023-06-19T06%3A46%3A57Z). I think this has `O(n)` complexity and is optimal so I don't see much benefit of implementing a heuristic in Python which will be both slower and worse (and more work)?

The main reason for not implementing optimal SPLIT right now is that I don't have enough time/motivation to understand the algorithm. I _could_ copy-paste the code from HGS-CVRP/HGS-VRPTW, but that will make it hard to bugfix and write tests for it.

The implemented SPLIT algorithms are not optimal anyway for VRPTW, so I hope that a sequential/naive-SPLIT is not too bad. If it is, then we can always decide to implement optimal SPLIT.

**Comment by N-Wouda** [comment]
In #235 I wrote:

>I'd like to avoid us having to bring back the whole SPLIT algorithm and decoding/encoding a giant tour, since that adds a lot of complexity for IMO little benefit.

That comment is relevant here. Do we still want to bring back the SPLIT algorithm, and all the complexity that entails? I see more value in @wouterkool's comment in PR #423:

>One thing I'd like to discuss though is what we think of the case where we have a VRP with > 1 vehicles where all clients can be served by a single vehicle. While not very common this is perfectly fine and can happen in practice. In this case we will get stuck without a warning.
>
>One idea could be to just have a crossover operator (pure python) that dynamically dispatches to OX or SREX based on the number of routes in the parent solution.

That would use our OX for TSP crossover dynamically depending on the number of vehicles in the current solution(s). We might want to consider the details of such a hybrid crossover here, rather than try to bring back OX for VRP / SPLIT.

**Comment by leonlan** [comment]
For reference, OX was implemented in #423 because SREX doesn't work for instances with a single route. But OX's use is restricted to TSP-like instances, because using it for VRP requires the SPLIT algorithm.

> I'd like to avoid us having to bring back the whole SPLIT algorithm and decoding/encoding a giant tour, since that adds a lot of complexity for IMO little benefit.

I agree with this. I recall that I stopped working on this (in August?) because SPLIT was non-trivial to implement. 

I think that Wouter's suggestion is good for handling the mentioned case. When both parents have 1 route, then using OX is the way to go. Otherwise, SREX should work fine right? It's a bit of a niche case but since the solution is rather easy we can implement it.
